lib/auth/authentication_backend.rb

```diff
@@ -0,0 +1,110 @@
+module AuthenticationBackend
+  # Abstracted Exceptions
+  # Aws::CognitoIdentityProvider::Errors::NotAuthorizedException,
+  # Aws::CognitoIdentityProvider::Errors::UserNotFoundException,
+  # Aws::CognitoIdentityProvider::Errors::InvalidParameterException,
+  # Aws::CognitoIdentityProvider::Errors::CodeMismatchException
+  # Aws::CognitoIdentityProvider::Errors::ServiceError
+  # Aws::CognitoIdentityProvider::Errors::AliasExistsException,
+  # Aws::CognitoIdentityProvider::Errors::UsernameExistsException => e
+  # Aws::CognitoIdentityProvider::Errors::ResourceNotFoundException
+
+  class NotAuthorizedException < StandardError; end
+  class UserGroupNotFoundException < StandardError; end
+  class AuthenticationBackendException < StandardError; end
+
+  class AuthenticationResponse
+    attr_reader :params, :id_token, :access_token
+
+    def initialize(params:, cognito_response:)
+      @params = params
+      @id_token = cognito_response[:authentication_result][:id_token]
+      @access_token = cognito_response[:authentication_result][:access_token]
+    end
+  end
+
+  class ChallengeResponse
```

Comment:
not a big fan of having multiple classes in a single file... once we bring all the other cognito calls here, it will be a huge file...

Reply:
I can move the classes to their own files if you'd prefer?  Given how small they are I didn't see much point in having them on their own but happy to make that change.

Reply:
happy for it to be part of the next PR when we bring the other methods

Reply:
Give me a minute and they'll be part of this PR... I am going to put up a new commit with the response_type stuff in from the previous commit.